aws/resource_aws_emr_instance_group.go

```diff
@@ -294,18 +334,88 @@ func instanceGroupStateRefresh(conn *emr.EMR, clusterID, igID string) resource.S
 	}
 }
 
-func resourceAwsEMRInstanceGroupDelete(d *schema.ResourceData, meta interface{}) error {
-	log.Printf("[WARN] AWS EMR Instance Group does not support DELETE; resizing cluster to zero before removing from state")
-	conn := meta.(*AWSClient).emrconn
-	params := &emr.ModifyInstanceGroupsInput{
-		InstanceGroups: []*emr.InstanceGroupModifyConfig{
-			{
-				InstanceGroupId: aws.String(d.Id()),
-				InstanceCount:   aws.Int64(0),
-			},
-		},
+// Populates an emr.EbsConfiguration struct
+func readEmrEBSConfig(d *schema.ResourceData) *emr.EbsConfiguration {
+	result := &emr.EbsConfiguration{}
+	if v, ok := d.GetOk("ebs_optimized"); ok {
+		result.EbsOptimized = aws.Bool(v.(bool))
 	}
 
-	_, err := conn.ModifyInstanceGroups(params)
-	return err
+	ebsConfigs := make([]*emr.EbsBlockDeviceConfig, 0)
+	if rawConfig, ok := d.GetOk("ebs_config"); ok {
+		configList := rawConfig.(*schema.Set).List()
+		for _, config := range configList {
+			conf := config.(map[string]interface{})
+			ebs := &emr.EbsBlockDeviceConfig{}
+			volumeSpec := &emr.VolumeSpecification{
+				SizeInGB:   aws.Int64(int64(conf["size"].(int))),
+				VolumeType: aws.String(conf["type"].(string)),
+			}
+			if v, ok := conf["iops"].(int); ok && v != 0 {
+				volumeSpec.Iops = aws.Int64(int64(v))
+			}
+			if v, ok := conf["volumes_per_instance"].(int); ok && v != 0 {
+				ebs.VolumesPerInstance = aws.Int64(int64(v))
+			}
+			ebs.VolumeSpecification = volumeSpec
+			ebsConfigs = append(ebsConfigs, ebs)
+		}
+	}
+	result.EbsBlockDeviceConfigs = ebsConfigs
+	return result
+}
+
+func fetchAllEMRInstanceGroups(conn *emr.EMR, clusterID string) ([]*emr.InstanceGroup, error) {
+	req := &emr.ListInstanceGroupsInput{
+		ClusterId: aws.String(clusterID),
+	}
+
+	var groups []*emr.InstanceGroup
+	marker := aws.String("intitial")
+	for marker != nil {
+		log.Printf("[DEBUG] EMR Cluster Instance Marker: %s", *marker)
+		respGrps, errGrps := conn.ListInstanceGroups(req)
```

Comment:
We should prefer to use the [SDK provider paginated function `ListInstanceGroupsPages()`](https://docs.aws.amazon.com/sdk-for-go/api/service/emr/#EMR.ListInstanceGroupsPages) instead of implementing our own, e.g.

```go
input := &emr.ListInstanceGroupsInput{
	ClusterId: aws.String(clusterID),
}
var groups []*emr.InstanceGroup

err := conn.ListInstanceGroupsPages(input, func(page *emr.ListInstanceGroupsOutput, lastPage bool) {
	groups = append(groups, page.InstanceGroups...)

	return !lastPage
})

return groups, err
```